src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreBuildHandler.cs

```diff
@@ -0,0 +1,172 @@
+        /// The cookie associated to the the <see cref="IVsUpdateSolutionEvents5"/> events.
+        /// </summary>
+        private uint _updateSolutionEventsCookieEx;
+
+        private SolutionRestoreBuildHandler()
+        {
+        }
+
+        // A constructor utilized for running unit-tests
+        public SolutionRestoreBuildHandler(
+            INuGetLockService lockService,
+            ISettings settings,
+            ISolutionRestoreWorker restoreWorker,
+            IVsSolutionBuildManager3 buildManager)
+        {
+            Assumes.Present(lockService);
+            Assumes.Present(settings);
+            Assumes.Present(restoreWorker);
+            Assumes.Present(buildManager);
+
+            LockService = new Lazy<INuGetLockService>(() => lockService);
+            Settings = new Lazy<ISettings>(() => settings);
+            SolutionRestoreWorker = new Lazy<ISolutionRestoreWorker>(() => restoreWorker);
+
+            _solutionBuildManager = buildManager;
+        }
+
+        public void Dispose()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            if (_updateSolutionEventsCookieEx != VSCOOKIE_NIL)
+            {
+                ((IVsSolutionBuildManager)_solutionBuildManager).UnadviseUpdateSolutionEvents(_updateSolutionEventsCookieEx);
+                _updateSolutionEventsCookieEx = VSCOOKIE_NIL;
+            }
+        }
+
+        // A factory method invoked internally only
+        internal static async Task<IDisposable> InitializeAsync(Microsoft.VisualStudio.Shell.IAsyncServiceProvider serviceProvider)
+        {
+            Assumes.Present(serviceProvider);
+
+            var instance = new SolutionRestoreBuildHandler();
+
+            var componentModel = await serviceProvider.GetComponentModelAsync();
+            componentModel.DefaultCompositionService.SatisfyImportsOnce(instance);
+
+            await instance.SubscribeAsync(serviceProvider);
+
+            return instance;
+        }
+
+        private async SystemTask SubscribeAsync(Microsoft.VisualStudio.Shell.IAsyncServiceProvider serviceProvider)
+        {
+            // Don't use CPS thread helper because of RPS perf regression
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            _solutionBuildManager = await serviceProvider.GetServiceAsync<SVsSolutionBuildManager, IVsSolutionBuildManager3>();
+            Assumes.Present(_solutionBuildManager);
+
+            ((IVsSolutionBuildManager6)_solutionBuildManager).AdviseUpdateSolutionEventsEx(this, out _updateSolutionEventsCookieEx);
+        }
+
+        #region IVsUpdateSolutionEvents5
+
+        public void UpdateSolution_QueryDelayBuildAction(uint dwAction, out Guid pActivityId, out IVsTask pDelayTask)
+        {
+            pActivityId = GuidList.guidNuGetSBMEvents;
+            pDelayTask = SolutionRestoreWorker.Value.JoinableTaskFactory.RunAsyncAsVsTask(VsTaskRunContext.UIThreadBackgroundPriority, async (token) =>
```

Comment:
`RunAsyncAsVsTask` likely is not supported in test environment. To resolve that and also to fix potential [VSTHRD102](https://github.com/Microsoft/vs-threading/blob/master/doc/analyzers/VSTHRD102.md) create a separate async method for internal predicate and call it in tests.